# Выбор локации для скважины

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Введение" data-toc-modified-id="Введение-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Введение</a></span></li><li><span><a href="#Загрузка-и-подготовка-данных" data-toc-modified-id="Загрузка-и-подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка и подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обучение-и-проверка-модели" data-toc-modified-id="Обучение-и-проверка-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение и проверка модели</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Подготовка-к-расчёту-прибыли" data-toc-modified-id="Подготовка-к-расчёту-прибыли-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Подготовка к расчёту прибыли</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Расчет-прибыли-и-рисков" data-toc-modified-id="Расчет-прибыли-и-рисков-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Расчет прибыли и рисков</a></span><ul class="toc-item"><li><span><a href="#Расчёт-прибыли-по-выбранным-скважинам" data-toc-modified-id="Расчёт-прибыли-по-выбранным-скважинам-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Расчёт прибыли по выбранным скважинам</a></span></li><li><span><a href="#Учет-рисков" data-toc-modified-id="Учет-рисков-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Учет рисков</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Итоги-исследования" data-toc-modified-id="Итоги-исследования-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Итоги исследования</a></span></li></ul></div>

## Введение

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

**Исходные данные**: Вам предоставлены пробы нефти в трёх регионах.

Данные геологоразведки трёх регионов находятся в файлах:

* `/datasets/geo_data_0.csv`
* `/datasets/geo_data_1.csv`
* `/datasets/geo_data_2.csv`

Описание данных:

* `id` — уникальный идентификатор скважины;
* `f0, f1, f2` — три признака точек (неважно, что они означают, но сами признаки значимы);
* `product` — объём запасов в скважине (тыс. баррелей).

**Цель** - построить модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

**Задачи:**
1. Загрузить и подготовить данные;
2. Обучить и проверить модель для каждого региона;
3. Подготовить даннные для расчёта прибыли;
4. Написать функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели;
5. Посчитать риски и прибыль для каждого региона;
6. Подвести итоги исследования.

**Условия:**
* Для обучения модели подходит только линейная регрессия (остальные — недостаточно предсказуемые).
* При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
* Бюджет на разработку скважин в регионе — 10 млрд рублей.
* При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
* После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью.

Данные синтетические: детали контрактов и характеристики месторождений не разглашаются.

**План работы:**

1. Загрузите и подготовьте данные. Поясните порядок действий.


2. Обучите и проверьте модель для каждого региона:

    2.1. Разбейте данные на обучающую и валидационную выборки в соотношении 75:25.

    2.2. Обучите модель и сделайте предсказания на валидационной выборке.
    
    2.3. Сохраните предсказания и правильные ответы на валидационной выборке.
    
    2.4. Напечатайте на экране средний запас предсказанного сырья и RMSE модели.
    
    2.5. Проанализируйте результаты.


3. Подготовьтесь к расчёту прибыли:
        
    3.1. Все ключевые значения для расчётов сохраните в отдельных переменных.
    
    3.2. Рассчитайте достаточный объём сырья для безубыточной разработки новой скважины. Сравните полученный объём сырья со средним запасом в каждом регионе.
    
    3.3. Напишите выводы по этапу подготовки расчёта прибыли.


4. Напишите функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели:
        
    4.1. Выберите скважины с максимальными значениями предсказаний.
    
    4.2. Просуммируйте целевое значение объёма сырья, соответствующее этим предсказаниям.
    
    4.3. Рассчитайте прибыль для полученного объёма сырья.


5. Посчитайте риски и прибыль для каждого региона:
        
    5.1. Примените технику Bootstrap с 1000 выборок, чтобы найти распределение прибыли.
    
    5.2. Найдите среднюю прибыль, 95%-й доверительный интервал и риск убытков. Убыток — это отрицательная прибыль.
    
    5.3. Напишите выводы: предложите регион для разработки скважин и обоснуйте выбор.



## Загрузка и подготовка данных

In [1]:
# импортируем необходимые библиотеки
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np

In [2]:
# импортируем данные
df0 = pd.read_csv('/datasets/geo_data_0.csv')
df1 = pd.read_csv('/datasets/geo_data_1.csv')
df2 = pd.read_csv('/datasets/geo_data_2.csv')

df = [df0, df1, df2]

# добавим в каждую таблицу признак региона
for i in range(0,3):
    df[i]['region'] = i

# соединим таблицы
df = pd.concat([df0, df1, df2])

# выведем основную информацию по датафреймам на экран
display(df)
df.info()
display(df.describe())

id        f0        f1        f2     product  region
0      txEyH  0.705745 -0.497823  1.221170  105.280062       0
1      2acmU  1.334711 -0.340164  4.365080   73.037750       0
2      409Wp  1.022732  0.151990  1.419926   85.265647       0
3      iJLyR -0.032172  0.139033  2.978566  168.620776       0
4      Xdl7t  1.988431  0.155413  4.751769  154.036647       0
...      ...       ...       ...       ...         ...     ...
99995  4GxBu -1.777037  1.125220  6.263374  172.327046       2
99996  YKFjq -1.261523 -0.894828  2.524545  138.748846       2
99997  tKPY3 -1.199934 -2.957637  5.219411  157.080080       2
99998  nmxp2 -2.419896  2.417221 -5.548444   51.795253       2
99999  V9kWn -2.551421 -2.025625  6.090891  102.775767       2

[300000 rows x 6 columns]

<class 'pandas.core.frame.DataFrame'>
Int64Index: 300000 entries, 0 to 99999
Data columns (total 6 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       300000 non-null  object 
 1   f0       300000 non-null  float64
 2   f1       300000 non-null  float64
 3   f2       300000 non-null  float64
 4   product  300000 non-null  float64
 5   region   300000 non-null  int64  
dtypes: float64(4), int64(1), object(1)
memory usage: 16.0+ MB


f0             f1             f2        product  \
count  300000.000000  300000.000000  300000.000000  300000.000000   
mean        0.547913      -1.516172       2.497439      85.441667   
std         5.316634       3.900220       2.916502      46.519494   
min       -31.609576     -26.358598     -12.088328       0.000000   
25%        -1.003037      -2.478939       0.648213      52.666629   
50%         0.347934      -0.229632       2.491215      84.038886   
75%         1.755365       0.678562       4.344943     124.174086   
max        29.421755      18.734063      16.739402     190.029838   

              region  
count  300000.000000  
mean        1.000000  
std         0.816498  
min         0.000000  
25%         0.000000  
50%         1.000000  
75%         2.000000  
max         2.000000

* проверка на дубликаты

In [3]:
# проверим поле id на наличие дубликатов
print('Дубликатов в поле "id" - ', df.duplicated('id').sum())

# удалим строки с дубликатам в столбце id
df.drop_duplicates(subset='id', inplace=True)

# проверим
print('Дубликатов в поле "id" после очистки - ', df.duplicated('id').sum())

Дубликатов в поле "id" -  49
Дубликатов в поле "id" после очистки -  0


In [4]:
# заменим номера индексов датафрейма на id скважин
df.set_index('id', inplace=True)
df

f0        f1        f2     product  region
id                                                     
txEyH  0.705745 -0.497823  1.221170  105.280062       0
2acmU  1.334711 -0.340164  4.365080   73.037750       0
409Wp  1.022732  0.151990  1.419926   85.265647       0
iJLyR -0.032172  0.139033  2.978566  168.620776       0
Xdl7t  1.988431  0.155413  4.751769  154.036647       0
...         ...       ...       ...         ...     ...
4GxBu -1.777037  1.125220  6.263374  172.327046       2
YKFjq -1.261523 -0.894828  2.524545  138.748846       2
tKPY3 -1.199934 -2.957637  5.219411  157.080080       2
nmxp2 -2.419896  2.417221 -5.548444   51.795253       2
V9kWn -2.551421 -2.025625  6.090891  102.775767       2

[299951 rows x 5 columns]

In [5]:
# проверим поле 'product' на наличие дубликатов
for i in range(3):
    dupl = df[df['region'] == i].duplicated('product').sum()
    print(f'В регионе {i} поле "product" содержит дубликаты в количестве: {dupl}')

В регионе 0 поле "product" содержит дубликаты в количестве: 0
В регионе 1 поле "product" содержит дубликаты в количестве: 99974
В регионе 2 поле "product" содержит дубликаты в количестве: 0


### Вывод

* данные по регионам собраны в единый датафрейм для удобства работы;
* пустые значения отсутcтвуют, данные имеют корректный формат;
* дубликаты в столбце `id` удалены;
* в регионе 2 практически все скважины имеют одинаковую добычу! В реальной жизни каждая скважина по добыче нефти уникальна и не может повторяться, особенно, что касается уровня добычи. Говорю как человек, который закончил профильный нефтегазовый ВУЗ :D 
* в признаках имеется значительный разброс.

## Обучение и проверка модели

Обучите и проверьте модель для каждого региона:
* Разбейте данные на обучающую и валидационную выборки в соотношении 75:25.
* Обучите модель и сделайте предсказания на валидационной выборке.
* Сохраните предсказания и правильные ответы на валидационной выборке.
* Напечатайте на экране средний запас предсказанного сырья и RMSE модели.
* Проанализируйте результаты.

In [6]:
# создадим словарь, в который запишем предсказания по регионам
pred_dict = {0:[],
             1:[],
             2:[]}


# циклом пройдем по датафреймам каждого региона
for i in range(3):
    
    region = df[df['region'] == i]
    # разделим признаки и целевой признак
    features = region.drop('product', axis=1)
    target = region['product']
    
    # разделим признаки и целевой признак на обучающий и валидационный набор
    features_train, features_valid, target_train, target_valid = train_test_split(features,
                                                                                  target,
                                                                                  test_size=0.25,
                                                                                  random_state=12345)
    # объекту model присвоим структуру данных LinearRegression
    model = LinearRegression()
    
    # обучим модель на обучающей выборке
    model.fit(features_train, target_train)
    
    # выполним предсказания на валидационной выборке
    predictions = model.predict(features_valid)
    
    # переведем список в структуру данных Series
    predictions = pd.Series(predictions,
                            index=features_valid.index)
    # запишем в словарь
    pred_dict[i] = predictions
    
    # рассчитаем rmse модели
    mse = mean_squared_error(pd.Series(target_valid), predictions)
    rmse = mse ** 0.5
    
    print(f'Средний объём запасов предсказанного сырья в 1 скважине для региона {i} - {predictions.mean():.2f} тыс. барр.')
    print(f'RMSE модели - {rmse:0.2f}')
    print()    

Средний объём запасов предсказанного сырья в 1 скважине для региона 0 - 92.79 тыс. барр.
RMSE модели - 37.85

Средний объём запасов предсказанного сырья в 1 скважине для региона 1 - 68.89 тыс. барр.
RMSE модели - 0.89

Средний объём запасов предсказанного сырья в 1 скважине для региона 2 - 95.10 тыс. барр.
RMSE модели - 39.94



### Вывод

1. Для каждого региона:
* выборки разделены на обучающий и валидационный наборы;
* выполнены предсказания среднего объёма запасов предсказанного сырья в скважине;
* рассчитана метрика качества `RMSE` для модели линейной регрессии.

2. Анализ показал:
    
    2.1 Регион 1:
    * `Средний объём запасов` предсказанного сырья в 1 скважине для региона 0 - `92.59 тыс. барр.`
    * `RMSE` модели - `37.58`
    
    2.2 Регион 2:
    * `Средний объём запасов` предсказанного сырья в 1 скважине для региона 1 - `68.73 тыс. барр.`
    * `RMSE` модели - `0.89`
    
    2.3 Регион 3:
    * `Средний объём запасов` предсказанного сырья в 1 скважине для региона 2 - `94.97 тыс. барр.`
    * `RMSE` модели - `40.03`


3. Наиболее точно модель линейной регрессии предсказала значения сырья в скважинах для региона №1, т.к. значение RMSE модели составило 0.89 тыс. барр. Предсказания по модели для регионов 0 и 2 показали результаты хуже - RMSE 37.58 и 40.03 тыс. барр. соответственно, что составляет практически половину от среднего запаса сырья в скважине этих регионов.

## Подготовка к расчёту прибыли

* Все ключевые значения для расчётов сохраните в отдельных переменных.
* Рассчитайте достаточный объём сырья для безубыточной разработки новой скважины. Сравните полученный объём сырья со средним запасом в каждом регионе.
* Напишите выводы по этапу подготовки расчёта прибыли.

In [7]:
# Бюджет на разработку скважин в регионе — 10 млрд рублей.
Budget = 10 ** 10
# При нынешних ценах один баррель сырья приносит 450 рублей дохода.
# При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
Count = 200

Revenue_per_barrel = 450

# рассчитаем необходимый объем баррелей для достижения безубыточной добычи
breakeven_point = Budget / Count / 450 / 1000
print(f'Минимальный объем добычи сырья для достижения точки безубыточности 1 скважины - {breakeven_point:.2f} тыс. барр./скважину')

Минимальный объем добычи сырья для достижения точки безубыточности 1 скважины - 111.11 тыс. барр./скважину


### Вывод

* Был рассчитан `минимальный объем добычи сырья для достижения точки безубыточности 1 скважины`, который составил - `111.11 тыс. барр./скважину`
* Точка безубыточной добычи 1 скважины выше, чем средний объем добычи на 1 скважине в рассматриваемых регионах, следовательно, предварительно можно предположить, что добыча в рассматриваемых регионах может быть убыточна.

## Расчет прибыли и рисков 

### Расчёт прибыли по выбранным скважинам

Напишите функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели:
* Выберите скважины с максимальными значениями предсказаний.
* Просуммируйте целевое значение объёма сырья, соответствующее этим предсказаниям.
* Рассчитайте прибыль для полученного объёма сырья.

In [8]:
# напишем функцию для расчета прибыли
def profit(data, predictions, Budget, rev_per_barrel):
    
    # Выберем скважины с максимальными значениями предсказаний.
    predictions_sorted = predictions.sort_values(ascending=False)[:200]

    # Просуммируйте целевое значение объёма сырья (колонка 'production'), 
    # соответствующее предсказаниям по ТОП-200 скважинам
    total_recovery = data['product'][predictions_sorted.index].sum()
        
    # Рассчитайте прибыль для полученного объёма сырья.
    net_profit = (total_recovery * rev_per_barrel * 1000 - Budget) / 10**9
    
    return net_profit, total_recovery

# Циклом пройдем по предсказанным данным добычи каждого региона
for i in range(3):
    net_profit, total_recovery = profit(df[df['region'] == i], # отфильтруем данные по каждому региону
                                        pred_dict[i],          # из pred_dict возьмем предсказания по каждому региону
                                        Budget,                # общий бюджет на разработку 200 скважин
                                        450)                   # выручка при продаже барреля нефти
    
    
    # Выведем информацию на экран
    print(f'Общий объем добычи по 200 скважинам в регионе {i} - {total_recovery:.2f} тыс. барр., чистая прибыль - {net_profit:.2f} млрд. руб.')

Общий объем добычи по 200 скважинам в регионе 0 - 29700.42 тыс. барр., чистая прибыль - 3.37 млрд. руб.
Общий объем добычи по 200 скважинам в регионе 1 - 27589.08 тыс. барр., чистая прибыль - 2.42 млрд. руб.
Общий объем добычи по 200 скважинам в регионе 2 - 27557.78 тыс. барр., чистая прибыль - 2.40 млрд. руб.


### Учет рисков

* Примените технику Bootstrap с 1000 выборок, чтобы найти распределение прибыли.
* Найдите среднюю прибыль, 95%-й доверительный интервал и риск убытков. Убыток — это отрицательная прибыль.
* Напишите выводы: предложите регион для разработки скважин и обоснуйте выбор.

Для применения техники Bootstrap с 1000 выборок выполним следующее:
1. Предварительная подготовка;
2. Напишем два цикла, в котором "пройдем" по информации датафреймов каждого региона;
3. По каждому региону в ходе 1000 итераций вложенного цикла:
* выполним 'train_test_split';
* обучим модель линейной регрессии; 
* выполним предсказания на валидационной выборке;
* рассчитаем RMSE модели, запишем в список словаря;
* полученные значения предсказаний передадим функции 'profit', которая рассчитает чистую прибыль и добычу по выборке из ТОП-200 по добыче нефти скважин. Значения запишем в список словаря.

In [9]:
# 1-шаг Предварительная подготовка
# Создадим словари, куда запишем данные rmse, чистой прибыли и добычи по каждому региону
rmse_dict = {0:list(),
             1:list(),
             2:list()}    

profit_dict = {0:list(),
               1:list(),
               2:list()}

total_recovery_dict = {0:list(),
                       1:list(),
                       2:list()}

In [11]:
%%time

# для воспроизводимости псевдослучайной последовательности применим np.random.seed(1000) с выборкой из 1000 псевдослучайных чисел
rng = np.random.seed(1000)

# 2-шаг Напишем два цикла, в котором "пройдем" по информации датафреймов каждого региона
for region in range(3):
    
    data = df[df['region'] == region]    
    
    # разделим признаки и целевой признак
    features = data.drop('product', axis=1)
    target   = data['product']
                
    # разделим признаки и целевой признак на обучающий и валидационный набор в пропорции 75:25 c помощью 'train_test_split'
    features_train, features_valid, target_train, target_valid = train_test_split(features,
                                                                                  target,
                                                                                  test_size=0.25,
                                                                                  random_state=12345)
    # объекту model присвоим структуру данных LinearRegression
    model = LinearRegression()
        
    # обучим модель на обучающей выборке
    model.fit(features_train, target_train)
        
# 3-шаг Техникой BOOTSTRAP найдем распределение прибыли и прочих необходимых данных по объектам валидационного набора
    for i in range(1000):
                
        # соберем новую выборку из 500 объектов валидационного набора
        features_valid_sampled = features_valid.sample(n=500,
                                                       replace=True,
                                                       random_state=rng)

        target_valid_sampled = target_valid.sample(n=500,
                                                   replace=True,
                                                   random_state=rng)
        
        target_valid_sampled = pd.Series(target_valid_sampled, 
                                         index=target_valid_sampled.index)

        
        # выполним предсказания на валидационной выборке
        pred = pd.Series(model.predict(features_valid_sampled),
                         index=features_valid_sampled.index)
               
        # рассчитаем RMSE модели и добавим в словарь
        mse = mean_squared_error(target_valid_sampled, pred)
        rmse_dict[region].append(mse ** 0.5)
        
        # передадим функции 'profit' предсказания на валидационной выборке
        net_profit, total_recovery = profit(data,                  # отфильтруем данные по каждому региону
                                            pred,                  # из pred_dict возьмем предсказания по каждому региону
                                            Budget,                # общий бюджет на разработку 200 скважин
                                            Revenue_per_barrel)    # выручка при продаже барреля нефти
        
        # запишем полученные значения в словари 'profit_dict','total_recovery_dict'
        profit_dict[region].append(net_profit)
        total_recovery_dict[region].append(total_recovery)
    
    print('Успех!')

Успех!
Успех!
Успех!
CPU times: user 16.3 s, sys: 46.6 ms, total: 16.4 s
Wall time: 17.9 s


* Найдите среднюю прибыль, 95%-й доверительный интервал и риск убытков. Убыток — это отрицательная прибыль.

In [13]:
# циклом пройдем по спискам словарей, в которые записаны данные чистой прибыли, добычи, ст.отклонения
for i in range(3):
    print(f'Параметры разработки ТОП-скважин региона {i}:')
    print(f' * Средний объем добычи нефти в скважине составит - {pd.Series(total_recovery_dict[i]).mean()/200:.2f} тыс. барр')
    print(f' * Средняя прибыль по добыче на 200 скважинах - {pd.Series(profit_dict[i]).mean():.2f} млрд руб.')
    print(f' * 95% доверительный интервал получения чистой прибыли лежит в диапазоне от {pd.Series(profit_dict[i]).quantile(0.025):.2f} до {pd.Series(profit_dict[i]).quantile(0.975):.2f} млрд руб.')
    print(f' * Риск убытков - {(pd.Series(profit_dict[i]) < 0).sum()/1000:.2%}')
    print()


Параметры разработки ТОП-скважин региона 0:
 * Средний объем добычи нефти в скважине составит - 115.57 тыс. барр
 * Средняя прибыль по добыче на 200 скважинах - 0.40 млрд руб.
 * 95% доверительный интервал получения чистой прибыли лежит в диапазоне от -0.10 до 0.90 млрд руб.
 * Риск убытков - 5.90%

Параметры разработки ТОП-скважин региона 1:
 * Средний объем добычи нефти в скважине составит - 116.40 тыс. барр
 * Средняя прибыль по добыче на 200 скважинах - 0.48 млрд руб.
 * 95% доверительный интервал получения чистой прибыли лежит в диапазоне от 0.08 до 0.87 млрд руб.
 * Риск убытков - 0.80%

Параметры разработки ТОП-скважин региона 2:
 * Средний объем добычи нефти в скважине составит - 115.24 тыс. барр
 * Средняя прибыль по добыче на 200 скважинах - 0.37 млрд руб.
 * 95% доверительный интервал получения чистой прибыли лежит в диапазоне от -0.15 до 0.89 млрд руб.
 * Риск убытков - 7.60%



### Вывод

1. На этапе расчета прибыли и рисков была написана функция для расчёта прибыли по скважинам с максимальными значениями предсказаний добычи и рассчитана сама прибыль для полученного объёма сырья:
  
  - Общий объем добычи по 200 скважинам в регионе 0 - 29700.42 тыс. барр., чистая прибыль - 3.37 млрд. руб.
  - Общий объем добычи по 200 скважинам в регионе 1 - 27589.08 тыс. барр., чистая прибыль - 2.42 млрд. руб.
  - Общий объем добычи по 200 скважинам в регионе 2 - 27557.78 тыс. барр., чистая прибыль - 2.40 млрд. руб.


2. С помощью техники Bootstrap с 1000 выборок найдено распределение прибыли по каждому региону, рассчитан 95% доверительный интервал, средний объем добычи, приходящийся на 1 скважину. `Анализ риска убытков показал, что только регион 1 оказался безубыточным`.


3. Рекомендуемым регионом для освоения является `Регион 1`, т.к. обладает:
  - `наибольшей средней прибылью при добыче на 200 скважинах - 0.48 млрд руб.`
  - `средний объем добычи нефти в скважине составит - 116.40 тыс. барр`, что выше точки безубыточной добычи в 111 тыс. барр./скважину;
  - `95% доверительный интервал получения чистой прибыли` лежит в диапазоне `от 0.08 до 0.87 млрд руб.`
  - `Риск убытков - 0.80%`

## Итоги исследования

`Выбор локации для скважины` - исследовательский проект, цель которого - построить модель машинного обучения для определения региона добычи с наибольшей прибылью. Для достижения указанной цели были поставлены и решены следующие задачи:

1. На первом этапе были загружены данные по регионам и собраны в единый датафрейм для удобства работы, проведена проверка на пустоты, корректный формат и дубликаты. Также были выявлены повторы в столбце 'id', а в регионе 2 практически все скважины имеют одинаковую добычу, что в действительности не может быть в принципе, т.к. каждая скважина по-своему уникальна как в плане геологических характеристик, параметров флюидов, так и плане профиля добычи;


2. На этапе "Обучения и проверки модели": 
    
    2.1. Для каждого региона:
    * выборки разделены на обучающий и валидационный наборы;
    * выполнены предсказания среднего объёма запасов предсказанного сырья в скважине;
    * рассчитана метрика качества `RMSE` для модели линейной регрессии.

    2.2. Анализ показал:
    
        Регион 1:
        * `Средний объём запасов` предсказанного сырья в 1 скважине для региона 0 - `92.59 тыс. барр.`
        * `RMSE` модели - `37.58`
    
        Регион 2:
        * `Средний объём запасов` предсказанного сырья в 1 скважине для региона 1 - `68.73 тыс. барр.`
        * `RMSE` модели - `0.89`
    
        Регион 3:
        * `Средний объём запасов` предсказанного сырья в 1 скважине для региона 2 - `94.97 тыс. барр.`
        * `RMSE` модели - `40.03`


    2.3. Наиболее точно модель линейной регрессии предсказала значения сырья в скважинах для региона №1, т.к. значение RMSE модели составило 0.89 тыс. барр. Предсказания по модели для регионов 0 и 2 показали результаты хуже - RMSE 37.58 и 40.03 тыс. барр. соответственно, что составляет практически половину от среднего запаса сырья в скважине этих регионов.
    

3. Далее была выполнена подготовка к расчету прибыли и рисков, которая заключалась в:
* рассчете `минимального объема добычи сырья для достижения точки безубыточности 1 скважины`, который составил - `111.11 тыс. барр./скважину`
* было отмечено, что `точка безубыточной добычи 1 скважины выше, чем средний объем добычи на 1 скважине в рассматриваемых регионах`, следовательно, предварительно можно предположить, что `добыча` в рассматриваемых регионах `может быть убыточна`.


4. На этапе расчета прибыли и рисков была написана функция для расчёта прибыли по скважинам с максимальными значениями предсказаний добычи и рассчитана сама прибыль для полученного объёма сырья:
  
* `Общий объем добычи по 200 скважинам в регионе 0 - 29700.42 тыс. барр., чистая прибыль - 3.37 млрд. руб.`
* `Общий объем добычи по 200 скважинам в регионе 1 - 27589.08 тыс. барр., чистая прибыль - 2.42 млрд. руб.`
* `Общий объем добычи по 200 скважинам в регионе 2 - 27557.78 тыс. барр., чистая прибыль - 2.40 млрд. руб.`


С помощью техники Bootstrap с 1000 выборок найдено распределение прибыли по каждому региону, рассчитан 95% доверительный интервал, средний объем добычи, приходящийся на 1 скважину. `Анализ риска убытков показал, что только регион 1 оказался безубыточным`.


Рекомендуемым регионом для освоения является `Регион 1`, т.к. обладает:

* `наибольшей средней прибылью при добыче на 200 скважинах - 0.48 млрд руб.`
* `средний объем добычи нефти в скважине составит - 116.40 тыс. барр`, что выше точки безубыточной добычи в 111 тыс. барр./скважину;
* `95% доверительный интервал получения чистой прибыли` лежит в диапазоне `от 0.08 до 0.87 млрд руб.`
* `Риск убытков - 0.80%`